src/index_map.rs

```diff
@@ -0,0 +1,432 @@
+    /// inserted, last in order, and `None` is returned.
+    /// ```
+    /// use nonempty_collections::*;
+    ///
+    /// let mut m = ne_indexmap!{"Duke" => "Leto", "Doctor" => "Yueh"};
+    /// assert_eq!(None, m.insert("Lady", "Jessica"));
+    /// assert_eq!(vec!["Duke", "Doctor", "Lady"], m.keys().copied().collect::<Vec<_>>());
+    ///
+    /// // Spoiler alert: there is a different duke at some point
+    /// assert_eq!(Some("Leto"), m.insert("Duke", "Paul"));
+    /// assert_eq!(vec!["Paul", "Yueh", "Jessica"], m.values().copied().collect::<Vec<_>>());
+    /// ```
+    pub fn insert(&mut self, k: K, v: V) -> Option<V> {
+        if k.equivalent(&self.head_key) {
+            Some(std::mem::replace(&mut self.head_val, v))
+        } else {
+            self.tail.insert(k, v)
+        }
+    }
+
+    /// Shrink the capacity of the map as much as possible.
+    pub fn shrink_to_fit(&mut self) {
+        self.tail.shrink_to_fit();
+    }
+
+    /// See [`IndexMap::with_capacity_and_hasher`].
+    pub fn with_capacity_and_hasher(capacity: usize, hasher: S, k: K, v: V) -> NEIndexMap<K, V, S> {
+        Self {
+            head_key: k,
+            head_val: v,
+            tail: IndexMap::with_capacity_and_hasher(capacity, hasher),
+        }
+    }
+
+    /// See [`IndexMap::with_hasher`].
+    pub fn with_hasher(hasher: S, k: K, v: V) -> NEIndexMap<K, V, S> {
+        Self {
+            head_key: k,
+            head_val: v,
+            tail: IndexMap::with_hasher(hasher),
+        }
+    }
+
+    /// Swaps the position of two key-value pairs in the map.
+    ///
+    /// # Panics
+    /// If `a` or `b` are out of bounds.
+    #[allow(clippy::panic)]
+    pub fn swap_indices(&mut self, mut a: usize, mut b: usize) {
+        if a == b && a < self.len() {
+            return;
+        }
+        if b == 0 {
+            b = a;
+            a = 0;
+        }
+
+        if a == 0 {
+            b -= 1;
+            let (existing_k, existing_v) = self
+                .tail
+                .swap_remove_index(b)
+                .unwrap_or_else(|| panic!("Index out of bounds {b}"));
+
+            let old_head_key = mem::replace(&mut self.head_key, existing_k);
+            let old_head_value = mem::replace(&mut self.head_val, existing_v);
+
+            let (index, _) = self.tail.insert_full(old_head_key, old_head_value);
+            // swap the newly inserted item to the original position
+            self.tail.swap_indices(index, b);
+        } else {
+            self.tail.swap_indices(a - 1, b - 1);
+        }
+    }
+}
+
+impl<K, V, S> PartialEq for NEIndexMap<K, V, S>
+where
+    K: Eq + Hash,
+    V: Eq,
+    S: BuildHasher,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.iter()
+            .all(|(k, v)| other.get(k).is_some_and(|ov| v == ov))
+    }
+}
+
+impl<K, V, S> Eq for NEIndexMap<K, V, S>
+where
+    K: Eq + Hash,
+    V: Eq,
+    S: BuildHasher,
+{
+}
```

Comment:
If empty, can be left out.

Reply:
`Eq` is a marker trait so I think we need to keep it here.

Reply:
Pardon me I didn't notice that it was `Eq`, my eyes had read a plain `impl` block for `NEIndexMap` with various contraints.